pyomo/util/slices.py

```diff
@@ -0,0 +1,289 @@
+
+def get_location_set_map(index, index_set):
+    """Map each value in an index to the set from which it originates
+
+    This function iterates over the "subsets" of `index_set` in the
+    forward direction, assigning sets to each value in `index`, until
+    it finds a set of dimension `None`. Then it iterates over the
+    reversed list of subsets, assigning sets, until it encounters
+    the same set of dimension `None`. All remaining values are assigned
+    to the set of dimension `None`. If a second such set is found,
+    an error is raised.
+
+    Parameters:
+    -----------
+    index: `tuple` or hashable scalar
+        The index whose values will be assigned to sets
+    index_set: `pyomo.core.base.set.SetProduct` or `pyomo.core.base.set.Set`
+        The index set from which `index` originates
+
+    Returns:
+    --------
+    `dict`: Maps the "locations" (indices) within the index to the set
+        from which it originates.
+
+    """
+    index = tuple(_to_iterable(index))
+    len_index = len(index)
+    locations_left = set(range(len_index))
+    location_set_map = {}
+
+    # Determine if we should exit early
+    if index_set is UnindexedComponent_set:
+        # index == (None,) and dimen is None in this case,
+        # so without this catch, the function would return
+        # {0: None}, which is not what we want for an
+        # unindexed component.
+        return {0: UnindexedComponent_set}
+    elif not normalize_index.flatten:
+        raise ValueError(
+            'get_location_set_map does not support the case where '
+            'normalize_index.flatten is False.'
+            # Although in this case, the location of an index should
+            # just be its position in the subsets list, so maybe
+            # the info we need is actually more simple to obtain.
+            )
+
+    subsets = list(_generate_subsets(index_set))
+
+    n_subsets = len(subsets)
+
+    location = 0
+    dimen_none_set = None
+    dimen_none_set_coord = None
+    # Step through subsets in forward order to assign as many
+    # locations as possible.
+    for sub_coord, sub in enumerate(subsets):
+        dimen = sub.dimen
+        if dimen is None:
+            dimen_none_set = sub
+            dimen_none_set_coord = sub_coord
+            break
+        for i in range(dimen):
+            location_set_map[location+i] = sub
+            locations_left.remove(location+i)
+        location += dimen
+
+    # We are either done or have encountered a set of dimen None
+    if not locations_left:
+        return location_set_map
+
+    location = len_index - 1
+    # Step through subsets in reverse order, assigning locations,
+    # until a set of dimen==None is encountered.
+    for sub_coord, sub in enumerate(reversed(subsets)):
+        sub_coord = n_subsets - sub_coord - 1
+        dimen = sub.dimen
+        if dimen is None:
+            if (dimen_none_set is not sub or
+                    dimen_none_set_coord != sub_coord):
```

Comment:
Isn't it sufficient to just check `dimen_none_set_coord != sub_coord`, as `dimen_none_set is not sub` always implies the coordinates are different?

Reply:
But the same set could appear in multiple locations, and I want to count that as invalid. The test on lines test_slices.py:517-521 would not raise the desired error if I omitted the second check.

Reply:
Agreed.  My point was that I think you *only* need the second check (`dimen_none_set_coord != sub_coord`) and you don't need the first (`dimen_none_set is not sub`)

Reply:
D'oh yep you're right. Changed.